### PR TITLE
[Feature] Add `AsAreabrick` attribute

### PR DESCRIPTION
**Pull request by jdreesen**
## Changes in this pull request  
This allows easier custom Areabrick registration, as you don't have to tag them manually anymore. So when you want them to live in e.g. `App\Custom\Location\Areabricks` instead of `App\Document\Areabrick`, you would have to register and tag them one by one at the moment. With this, you can just register the namespace as a resource, like:

```yaml
# config/services.yaml
services: 
    App\Custom\Location\Areabricks\:
        resource: '../src/App/Custom/Location/Areabricks/*'
```

and then tag the class with the `AsAreabrick` attribute, like:

```php
<?php

namespace App\Custom\Location\Areabricks;

use Pimcore\Extension\Document\Areabrick\AreabrickInterface;
use Pimcore\Extension\Document\Areabrick\Attribute\AsAreabrick;
use Psr\Log\LoggerInterface;

#[AsAreabrick(id: 'iframe')]
class Iframe implements AreabrickInterface
{
    // ...
}
```

---

The attribute can also be used to manually define the brick's ID when using the auto-registration. In this case, it won't be auto-generated from the class name anymore.

### WHAT
copilot:summary

copilot:poem

### HOW
copilot:walkthrough

### CODE REVIEW

**Comment by passioneight** [comment]
Awesome, @jdreesen :-)

> The attribute can also be used to manually define the brick's ID when using the auto-registration. In this case, it won't be auto-generated from the class name anymore.

It looks like the ID _has to be_ defined when using the attribute. Maybe making the `id` optional would increase DX, because the devs do not _have to_ provide an ID for every area, but _can_ customize it if needed. Wdyt?

**Comment by jdreesen** [comment]
Right, at the moment the ID *must* be defined when using the attribute. That's because I was initially only thinking of an alternative to manually tagging the brick, where the ID *must* be defined as well. But we can probably remove that requirement from manually tagged bricks and use the same logic as for automatically registered bricks when the tag has no ID.

When using the attribute for automatically registered bricks, it would then be a no-op if there is no ID.

**Comment by dvesh3** [comment]
@jdreesen please rebase your PR to fix the failing tests. thanks!

**Comment by passioneight** [comment]
@jdreesen of course, the noop is unfortunate, but it might just pave the way for additional improvements. For example, the auto-registration, while convenient, adds complexity for new developers. Having only the `AsAreabrick` attribute will let the devs jump to any related code, making it easier to understand what's going on behind the scenes. 

Building on top of this, Pimcore could implement a little something to support inheritance of this very attribute (since attributes are not inherited by default). Which means the devs only have to extend an abstract class (e.g., `AbstractTemplateAreabrick`), but can still add the attribute in their code if they need a different ID.

So, eventually only areas with the `AsAreabrick` attribute should be auto-registered. This way, the noop is actually an _op_ again.

At least to me, this would increase DX significantly, as one can always directly find their way to the related code using their preferred IDE, while also preserving the auto-registration feature.

But then again, that's just some thoughts I had when I saw your PR, maybe a discussion would shed more light on this topic? Eagerly looking forward to your thoughts on this.

> As a side note, renaming the attribute to `AsArea` could be beneficial, since it's also called `pimcore_area` and `pimcore_areablock` rather than `pimcore_areabrick` and `pimcore_areabrickblock`, when rendering them in the template. Wdyt?

**Comment by jdreesen** [comment]
@dvesh3 done. Pipeline is green now :)

@passioneight 

> For example, the auto-registration, while convenient, adds complexity for new developers. Having only the AsAreabrick attribute will let the devs jump to any related code, making it easier to understand what's going on behind the scenes.

I also had these thoughts and agree that the attribute could replace auto-registration.

> Building on top of this, Pimcore could implement a little something to support inheritance of this very attribute (since attributes are not inherited by default). Which means the devs only have to extend an abstract class (e.g., AbstractTemplateAreabrick), but can still add the attribute in their code if they need a different ID.

I'm not sure if it's that easy to integrate attribute inheritance into Symfony's container magic, but we can try.

But apart from that, I don't think it will degrade the DX if developers have to add this attribute manually when creating an areabrick.

> As a side note, renaming the attribute to `AsArea` could be beneficial, since it's also called `pimcore_area` and `pimcore_areablock` rather than `pimcore_areabrick` and `pimcore_areabrickblock`, when rendering them in the template. Wdyt?

I'm fine with both. I think that's something the Pimcore core team should decide.

In general, I'd like to get some input from @pimcore on how they would like this feature to work and if they would be okay with deprecating the auto-registration of areabricks in favor of the new attribute.

**Comment by jdreesen** [comment]
I just made the ID in the attribute/tag optional. If it's not defined, it will be auto-generated from the class name, like it's done when auto-registering the bricks.

File: bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php
Comment (dvesh3) [inline bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php:134]: It doesn't work when using the attribute (with or without id) as defined in the description. getting this error:
image

**Comment by dvesh3** [comment]
If we really want to get away from the auto-registration, then I would suggest to add more properties in the attribute so that we are completely detached from the auto stuff like finding brick templates. so this would look like 
```
#[AsAreabrick(
    id: 'iframe',
    template: 'custom/location/areabricks/iframe.twig',
)]
```
and we deprecate the `TemplateAreabrickInterface` along with `TEMPLATE_LOCATION_` constants. wdyt?

**Comment by passioneight** [comment]
@dvesh3 I agree, it would be much better to have the template explicitly in the area. However, I like to think of an area as some kind of special controller. As such, I think it would be nicer to have the same options as for controllers. That is, the possibility to return the template in the area's `action` method, or leveraging the `Template` attribute.

Wdyt? :-)

**Comment by dvesh3** [comment]
yes, this could replace `action` & `postRenderAction` which at the moment relies on `Info` object to pass the params to the views.

**Comment by passioneight** [comment]
> yes, this could replace action [...]

Could you elaborate a bit further as to why the `action` would be replaced? :-)

---

I too would like to share more details about how I imagine an approximate implementation of an area:
```php
<?php

namespace App\Area;

use Symfony\Component\HttpFoundation\Request;

// Would be really cool to have the attribute in AbstractArea, but that probably requires a lot of work
#[AsArea]
class Wysiwyg extends AbstractArea
{
    // Either the template is provided via this attribute ...
    #[Template('area/wysiwyg.html.twig')]
    public function action(Request $request): array|Response
    {
        $foo = 'bar';

        // ... or the template is provided via the return value
        return new Response('area/wysiwyg.html.twig', [
            'foo' => $foo
        ]);
    }
}
```

> Not too sure about returning a `Response` though. But it I think it helps clarifying where I'm going with this :-)

I'd also expect a call like this to pass the values to the area:

```twig
{{ pimcore_area('wysiwyg', {
    'type': 'wysiwyg',
    'params': {
        'text': product.description
    }
}) }}
```

It is then the developers responsibility to support the passed parameter in either the action or the template like so:

```twig
{% set text = text is defined ? text : pimcore_wysiwyg('text') %}

{# "visible" is a test similar to this: !($value instanceof Editable ? $value->isEmpty() : empty($value)) #}
{% if editmode or text is visible %}
    {{ text|raw }}
{% endif %}
```

**Comment by dvesh3** [comment]
@jdreesen would you be interested in implementing the `template` property to replace auto brick templates and deprecating auto-registration in this PR? otherwise we can do this in a follow up 😊

**Comment by jdreesen** [comment]
Sure, just give me some time, as I'm very busy at the moment. If you don't wanna wait, you can merge this and I'll do a followup PR.

File: bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php
Comment (dvesh3) [reply bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php:134]: @jdreesen we are reviewing this PR now to merge it. Could you please confirm this issue? thanks!

Comment (jdreesen) [reply bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php:134]: Right, the last commit was broken. I fixed and force-pushed it. Should work now.

Note that I'm on vacation now for the next two weeks (until August 16).

Comment (dvesh3) [reply bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php:134]: @jdreesen thanks Jacob!! I'll review it soon. enjoy your vacation 😊

**Review by dvesh3 — APPROVED** [review]

**Comment by dvesh3** [comment]
@jdreesen works like a charm 🚀 thanks a lot!

**Pull request merged**
